Unable to build google-cloud-cpp storage library with GCC 13

**Does this issue affect the google-cloud-cpp project?** yes

**What component of `google-cloud-cpp` is this related to?** `google/cloud/storage`

The build with compiler GCC 13 might be affected for other components too.

**Describe the bug** Unable to build google-cloud-cpp storage library with GCC 13

```
/release/master/extensions/.build/io/cpp/thirdparty/gcs/google-cloud-cpp-v2.15.0/google/cloud/internal/oauth2_external_account_credentials.cc: In member function ‘virtual google::cloud::v2_15::StatusOr<google::cloud::v2_15::AccessToken> google::cloud::oauth2_internal::v2_15::ExternalAccountCredentials::GetToken(std::chrono::_V2::system_clock::time_point)’:
/release/master/extensions/.build/io/cpp/thirdparty/gcs/google-cloud-cpp-v2.15.0/google/cloud/internal/oauth2_external_account_credentials.cc:159:33: warning: redundant move in initialization [-Wredundant-move]
  159 |   if (!payload) return std::move(payload.status());
      |                        ~~~~~~~~~^~~~~~~~~~~~~~~~~~
/release/master/extensions/.build/io/cpp/thirdparty/gcs/google-cloud-cpp-v2.15.0/google/cloud/internal/oauth2_external_account_credentials.cc:159:33: note: remove ‘std::move’ call
In file included from /release/master/extensions/.build/io/cpp/thirdparty/gcs/google-cloud-cpp-v2.15.0/google/cloud/internal/openssl_util.cc:15:
/release/master/extensions/.build/io/cpp/thirdparty/gcs/google-cloud-cpp-v2.15.0/google/cloud/internal/openssl_util.h:35:27: error: ‘uint8_t’ is not a member of ‘std’; did you mean ‘wint_t’?
   35 | StatusOr<std::vector<std::uint8_t>> SignUsingSha256(
      |                           ^~~~~~~
      |                           wint_t
/release/master/extensions/.build/io/cpp/thirdparty/gcs/google-cloud-cpp-v2.15.0/google/cloud/internal/openssl_util.h:35:27: error: template argument 1 is invalid
/release/master/extensions/.build/io/cpp/thirdparty/gcs/google-cloud-cpp-v2.15.0/google/cloud/internal/openssl_util.h:35:27: error: template argument 2 is invalid
/release/master/extensions/.build/io/cpp/thirdparty/gcs/google-cloud-cpp-v2.15.0/google/cloud/internal/openssl_util.h:35:34: error: template argument 1 is invalid
   35 | StatusOr<std::vector<std::uint8_t>> SignUsingSha256(
      |                                  ^~
/release/master/extensions/.build/io/cpp/thirdparty/gcs/google-cloud-cpp-v2.15.0/google/cloud/internal/openssl_util.cc:77:37: error: ambiguating new declaration of ‘google::cloud::v2_15::StatusOr<std::vector<unsigned char> > google::cloud::v2_15::internal::SignUsingSha256(const std::string&, const std::string&)’
   77 | StatusOr<std::vector<std::uint8_t>> SignUsingSha256(
      |                                     ^~~~~~~~~~~~~~~
/release/master/extensions/.build/io/cpp/thirdparty/gcs/google-cloud-cpp-v2.15.0/google/cloud/internal/openssl_util.h:35:37: note: old declaration ‘int google::cloud::v2_15::internal::SignUsingSha256(const std::string&, const std::string&)’
   35 | StatusOr<std::vector<std::uint8_t>> SignUsingSha256(
      |                                     ^~~~~~~~~~~~~~~
gmake[5]: *** [google/cloud/CMakeFiles/google_cloud_cpp_rest_internal.dir/build.make:510: google/cloud/CMakeFiles/google_cloud_cpp_rest_internal.dir/internal/openssl_util.cc.o] Error 1
gmake[5]: *** Waiting for unfinished jobs....
gmake[4]: *** [CMakeFiles/Makefile2:288: google/cloud/CMakeFiles/google_cloud_cpp_rest_internal.dir/all] Error 2
gmake[3]: *** [Makefile:136: all] Error 2
```

**To Reproduce** Steps to reproduce the behavior:

1. Build google-cloud-cpp with compiler GCC 13

**Expected behavior** google-cloud-cpp should compile with compiler GCC 13

**Operating system:** 
```
NAME="Red Hat Enterprise Linux"
VERSION="9.2 (Plow)"
ID="rhel"
ID_LIKE="fedora"
VERSION_ID="9.2"
PLATFORM_ID="platform:el9"
PRETTY_NAME="Red Hat Enterprise Linux 9.2 (Plow)"
```
**What compiler and version are you using?** 
```
[root@7a55e5ef46b5 .build]# g++ -v
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/opt/tools/bin/../libexec/gcc/x86_64-pc-linux-gnu/13.1.0/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /home/jonathon/Work/tc/gcc-13.1.0/configure CXX= CC=/usr/bin/gcc --prefix=/home/jonathon/Work/tc/rocky-9-x86_64-13.1.0-main-3.1 --libdir=/home/jonathon/Work/tc/rocky-9-x86_64-13.1.0-main-3.1/lib --program-suffix=-13.1.0 --disable-multilib --with-system-zlib --enable-lto --enable-tls --enable-libmpx --enable-languages=c,c++
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 13.1.0 (GCC)
```
**What version of `google-cloud-cpp` are you using?** 
```
[root@7a55e5ef46b5 google-cloud-cpp-v2.15.0]# git rev-parse HEAD
ce74cb40a606c25f8d45ba9ac62d4433ba38ea2e
```
**Additional context** 
1. Refer https://gcc.gnu.org/gcc-13/porting_to.html

--- a/ci/cloudbuild/dockerfiles/demo-fedora.Dockerfile
+++ b/ci/cloudbuild/dockerfiles/demo-fedora.Dockerfile
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-FROM fedora:37
+FROM fedora:38
 ARG NCPU=4
 
 ## [BEGIN packaging.md]
@@ -25,7 +25,7 @@ RUN dnf makecache && \
         openssl-devel patch unzip tar wget zip zlib-devel
 # ```
 
-# Fedora 37 includes packages, with recent enough versions, for most of the
+# Fedora 38 includes packages, with recent enough versions, for most of the
 # direct dependencies of `google-cloud-cpp`.
 
 # ```bash
@@ -54,15 +54,12 @@ RUN curl -fsSL https://pkgconfig.freedesktop.org/releases/pkg-config-0.29.2.tar.
     ldconfig
 # ```
 
-# The version of RE2 included with this distro hard-codes C++11 in its
-# pkg-config file. This is fixed in later versions, and it is unnecessary as
-# Fedora's compiler defaults to C++17.  If you are not planning to use
-# `pkg-config(1)` you can ignore this step.  Alternatively, you can install
-# RE2 and gRPC from source.
-
-# ```bash
-WORKDIR /var/tmp/build/re2
-RUN sed -i 's/-std=c\+\+11 //' /usr/lib64/pkgconfig/re2.pc
+# Older versions of Fedora hard-code RE2 to use C++11. It was fixed starting
+# with Fedora:38. If you using Fedora >= 38 or you are not planning to use
+# `pkg-config(1)` you can ignore this step.  Alternatively, you can install RE2
+# and gRPC from source.
+# ```
+# sed -i 's/-std=c\+\+11 //' /usr/lib64/pkgconfig/re2.pc
 # ```
 
 # The following steps will install libraries and tools in `/usr/local`. By

--- a/ci/generate-markdown/generate-packaging.sh
+++ b/ci/generate-markdown/generate-packaging.sh
@@ -33,7 +33,7 @@ file="doc/packaging.md"
   # A "map" (comma separated) of dockerfile -> summary.
   DOCKER_DISTROS=(
     "demo-alpine-stable.Dockerfile,Alpine (Stable)"
-    "demo-fedora.Dockerfile,Fedora (37)"
+    "demo-fedora.Dockerfile,Fedora (38)"
     "demo-opensuse-leap.Dockerfile,openSUSE (Leap)"
     "demo-ubuntu-jammy.Dockerfile,Ubuntu (22.04 LTS - Jammy Jellyfish)"
     "demo-ubuntu-focal.Dockerfile,Ubuntu (20.04 LTS - Focal Fossa)"

--- a/doc/packaging.md
+++ b/doc/packaging.md
@@ -216,7 +216,7 @@ cmake --build cmake-out --target install
 </details>
 
 <details>
-<summary>Fedora (37)</summary>
+<summary>Fedora (38)</summary>
 <br>
 
 Install the minimal development tools:
@@ -227,7 +227,7 @@ sudo dnf install -y cmake curl findutils gcc-c++ git make ninja-build \
         openssl-devel patch unzip tar wget zip zlib-devel
 ```
 
-Fedora 37 includes packages, with recent enough versions, for most of the direct
+Fedora 38 includes packages, with recent enough versions, for most of the direct
 dependencies of `google-cloud-cpp`.
 
 ```bash
@@ -256,13 +256,12 @@ sudo make install && \
 sudo ldconfig
 ```
 
-The version of RE2 included with this distro hard-codes C++11 in its pkg-config
-file. This is fixed in later versions, and it is unnecessary as Fedora's
-compiler defaults to C++17. If you are not planning to use `pkg-config(1)` you
-can ignore this step. Alternatively, you can install RE2 and gRPC from source.
+Older versions of Fedora hard-code RE2 to use C++11. It was fixed starting with
+Fedora:38. If you using Fedora >= 38 or you are not planning to use
+`pkg-config(1)` you can ignore this step. Alternatively, you can install RE2 and
+gRPC from source.
 
-```bash
-mkdir -p $HOME/Downloads/re2 && cd $HOME/Downloads/re2
+```
 sed -i 's/-std=c\+\+11 //' /usr/lib64/pkgconfig/re2.pc
 ```
 

